cosmwasm/register-p2p-tokens.ts

```diff
@@ -0,0 +1,219 @@
+    // after the ITS contracts migration to `v2.2.0`.
+    needsAlignment?: boolean;
+};
+
+export type SquidToken = {
+    tokenId: string;
+    decimals: number;
+    tokenType: 'interchain' | 'customInterchain' | 'canonical';
+    chains: SquidTokenData[];
+    originAxelarChainId?: string;
+};
+
+export type SquidTokens = {
+    [tokenId: string]: SquidToken;
+};
+
+export type SquidTokenInfoFile = {
+    tokens: SquidTokens;
+};
+
+function getOriginChain(tokenData: SquidToken) {
+    // TODO tkulik: Why?
+    // If only a single chain is untracked, use that chain
+    const untracked = tokenData.chains.filter((chain) => !chain.trackSupply);
+    if (untracked.length === 1) {
+        printInfo(`Untracked token ${tokenData.tokenId} on ${untracked[0].axelarChainId}`);
+        return untracked[0].axelarChainId;
+    }
+
+    // Use ethereum as the origin chain if the token is deployed on any of the Ethereum chains.
+    const ethereumChains = ['ethereum', 'core-ethereum', 'ethereum-sepolia', 'core-ethereum-sepolia', 'eth-sepolia'];
+    const ethereumChain = tokenData.chains.find((chain) => ethereumChains.includes(chain.axelarChainId.toLowerCase()));
+    if (ethereumChain) {
+        return ethereumChain.axelarChainId;
+    }
+
+    // Use the first chain that shows up.
+    return tokenData.chains[0].axelarChainId;
+}
+
+async function getSupply(tokenAddress: string, rpc: string): Promise<string> {
+    const provider = getDefaultProvider(rpc);
+    const token = new Contract(tokenAddress, IInterchainToken.abi, provider);
+    const supply = await token.totalSupply();
+    return supply.toString();
+}
+
+async function forEachTokenInFile(
+    config: ConfigManager,
+    options,
+    processToken: (tokenData: SquidToken, tokenOnChain: SquidTokenData) => Promise<void>,
+) {
+    const { env, tokenIds, chains } = options;
+    const tokenIdsToProcess = new Set(tokenIds);
+    const chainsToProcess = new Set(chains);
+    const tokenInfoString = fs.readFileSync(`axelar-chains-config/info/tokens-p2p/tokens-${env}.json`, 'utf8');
+    const tokenInfo = JSON.parse(tokenInfoString) as SquidTokenInfoFile;
+    const interchainTokenServiceAddress = config.getContractConfig('InterchainTokenService').address;
+
+    if (!interchainTokenServiceAddress) {
+        throw new Error('InterchainTokenService contract address not found');
+    }
+
+    const promises = Object.values(tokenInfo.tokens)
+        .filter((tokenData: SquidToken) => (tokenIds ? tokenIdsToProcess.has(tokenData.tokenId) : true))
+        .flatMap((tokenData: SquidToken) => {
+            return tokenData.chains
+                .filter((chain: SquidTokenData) => {
+                    try {
+                        return (
+                            tokenData.tokenType === 'interchain' &&
+                            (chains ? chainsToProcess.has(chain.axelarChainId.toLowerCase()) : true) &&
+                            chain.axelarChainId !== tokenData.originAxelarChainId &&
```

Comment:
Squid config must have originAxelarChainId set, we should error if not. So there's no fallback logic needed